Changed Goal and Log level

theis999 opened the pull request:
Added dependency on MOD: Speedrun_goals
Added log config to control print level
Added print config for more specific printing control
Added variables.lua file to hold other generated variables 
Removed goal files as goals are now controlled by Speedrun_goals

Goals and log-level now have separate menus. Where the goal is controlled by speedrun goals.

### Speedrun goals
Speedrun goal handles rock yield and always day.

settings-updates.lua controls the settings of speedrun goals, so we can set the settings how we want them.

### Variables
The variables.lua file allows us to insert the variables we want / need without messing with control.lua or steps.lua. As both files are sufficiently large already.

### Log level
I have included 3 levels of logging. So we can reduce the amount of information printed to console more precisely.
We can easily enough introduce new levels with this setup.

In addition more specific settings are introduced to control printing of comment, research queue and save game.

MortenTobiasNielsen commented on Lua Files/control.lua:
```diff
@@ -115,13 +115,19 @@ local on_player_created = function(event)
 
 --Print message intended for viewers
-local function msg(msg)
+local function Message(msg)
```
Good idea to use message instead of msg - can you make the change to the input as well? :)

theis999 replied:
Done

MortenTobiasNielsen commented on Lua Files/control.lua:
```diff
@@ -1,9 +1,9 @@
-local debug_state = require("goal")
+local tas_generator = require("variables")
+local steps = require("steps")
+local debug_state = LOGLEVEL < 2
```
Why is debug_state active when LOGLEVEL is below 2 when it needs to be LOGLEVEL == 0 for it to print a debug message? (line 130)?

theis999 replied:
This might have been a stupid decision from me.
Previous we had 2 configurations debug and not_debug, while it is obvious that the new "Release" configuration match not_debug. It is however not obvious if debug matches the new "Debug" or "Development".

The best approach might be to just straight up remove debug_state entirely?

MortenTobiasNielsen replied:
I guess LOGLEVEL could be used instead. So when debug_state should be false we set the LOGLEVEL to 2?

theis999 replied:
I have revamped debug_state so it is exclusive to compatibility mode.

MortenTobiasNielsen commented on Lua Files/control.lua:
```diff
@@ -199,12 +205,12 @@ end
 	save_global()
 	if game.is_multiplayer() then
-		debug(string.format("Step: %s, saving game as %s", task, nameOfSaveGame))
+		if PRINT_SAVEGAME then debug(string.format("Step: %s, saving game as %s", task, nameOfSaveGame)) end
```
Why are LOGLEVEL, PRINT_SAVEGAME, etc. in all caps?

theis999 replied:
It was meant to indicate that they are global variables. But it does follow the Lua style for constants.

MortenTobiasNielsen commented on Lua Files/control.lua:
```diff
@@ -115,13 +115,19 @@ local on_player_created = function(event)
     player.print(msg)
 end
 
+local function Comment(msg)
```
Good idea to add a Comment function - cleans up the code nicely.

MortenTobiasNielsen commented on Lua Files/control.lua:
```diff
@@ -1772,19 +1777,6 @@ script.on_event(defines.events.on_player_mined_entity, function(event)
 		return
 	end
 
-	if event.entity.name == "rock-huge" then
-		mining_event_replace(event, "coal", 47)
-		mining_event_replace(event, "stone", 47)
-	end
-
-	if event.entity.name == "rock-big" then
-		-- do nothing, big rocks are always 20 stone
-	end
-
-	if event.entity.name == "sand-rock-big" then
-		mining_event_replace(event, "stone", 24)
-	end
-
```
I am not sure I like this idea that the tool is so tightly coupled with an external tool.

theis999 replied:
You do have mod privilegies so you can ensure that it is up to standard.
For release of a TAS the supporting mod is automatically downloaded & enabled, and for development it should be fairly easy for most people to ensure it is enabled.

Is there any job you would be fine with having a supporting mod take care of?

MortenTobiasNielsen replied:
I appreciate that I have mod privileges, but I would still like this tool to be able to function without the user having to "remember" that a mod is needed for it to function. 

I think the idea of having a nice GUI for the winning condition makes a lot of sense and will enhance the use experience, and it doesn't need to be a part of the FTG tool, so it is a good idea to externalize it, like the other semi-support tools.

theis999 replied:
It is not exactly something you can forget. As the TAS mod simply won't load without the dependency being enabled too.

MortenTobiasNielsen replied:
Fair, but it really is nicer that it just works, instead of it telling me that it doesn't work because I don't have a "third-party" mod installed.

theis999 replied:
I am not going to pretend i understand the issue you have. If the developer distribute their TAS through the mod portal dependencies are automatically download & enabled. If the developer have created other TAS's then they should know what to do. If the developer switch between different TAS mods, the mod will automatically be enabled. When is this an issue?

Seeing that the mod is not loaded is simple enough:
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/18309265/26ad5c7f-c724-4605-a84d-12cfeff0b615)

MortenTobiasNielsen replied:
How would the developer know that they suddenly have to install a mod to make it work - it haven't been needed before? I am aware that it is easy to do - like it is easy to download an extension, but I still would still like for a tool to at least function without external things. 

Why is it so important to tightly couple the goal tool instead of emitting an event like we do with other things?

MortenTobiasNielsen commented on Lua Files/settings-updates.lua:
```diff
@@ -0,0 +1,38 @@
+local settings = {
+    ["t-sg-speedrun-goal"] = {type = "string-setting", value = GOAL}, --["Steelaxe%", "GOTLAP", "Default settings", "Any%", "Robot rush"]
+    ["t-sg-always-day"] = {type = "bool-setting", value = true},--[true,false]
+    ["t-sg-rock-rng"] = {type = "string-setting", value = "Static"},--["Default", "Custom", "Static"]
+    ["t-sg-rock-rng-seed"] = {type = "int-setting", value = 5},--[1 : 100]
+    ["t-sg-rock-huge-coal"] = {type = "int-setting", value = 47},--[24 : 50]
+    ["t-sg-rock-huge-stone"] = {type = "int-setting", value = 47},--[24 : 50]
+    ["t-sg-sand-rock-big-stone"] = {type = "int-setting", value = 24},--[19 : 25]
```
I very much don't like this idea - I think we have talked about it at a few times.

theis999 replied:
I've removed the reliance on Speedrun goals' rock handling.
So it is only handles ending the game.

MortenTobiasNielsen reviewed (changes requested):
It would be nice if there was a little explanation of what each log level does - it is not really obvious what debug, development and release does.

theis999 commented:
It is a bit annoying that the help doesn't show up for log-levels as i did actually type it out.
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/18309265/589f93d3-400d-4743-b509-502b7798e308)

**Speedrun Goals**
One of my projects for this mod is to create a custom GUI for winning the game. Which includes the total time taken for beating the game in milliseconds and time to research technologies instead of kills
![image](https://github.com/MortenTobiasNielsen/Factorio-TAS-Generator/assets/18309265/9791e9c1-f80f-4ca9-a8a2-8128d7ecd56d)
As this is likely to include several lines of code to create, i want it far away from the rest of the TAS handling.

MortenTobiasNielsen commented:
Strange that the help message isn't shown, but seems to be an issue with the framework, so nothing really to do about it other than crossing fingers that it will be solved at some point. 

Sounds like a nice mod, which will be useful for many players. I did though think that speed runners aren't allowed to use mods, so will they be able to use this?

theis999 commented:
> ...I did though think that speed runners aren't allowed to use mods, so will they be able to use this?

They do, so they won't. But that doesn't particularly matter to me :)

I've changed the integration with the mod, so it only handles the endgame trigger and displaying the endgame gui.
I have removed most uses of debug_state.
I have added a new Error method, which should only be used for fatal errors.

MortenTobiasNielsen commented:
When I generate script with this branch my run doesn't work at all. It seems to just be something with it needing a dependency, but I would really appreciate it if that wasn't the case. I do not think the dependency provides enough value for it to warrant a breaking change.

theis999 commented:
I have run my steel axe TAS using this branch, and it worked perfectly fine. Providing me with End game GUI that i previously had to hack in.

theis999 committed:
Added debug levels.
Split debug from goal, old debug will default to steelaxe.
Merged all goals into one file.
Changed "Any %" to "Any%".
Added variables file to insert other variable data that isn't steps.
Moved generator_thumbprint to variables file.
Capitalized lua warning and debug function.

theis999 commented:
I'll close this PR and reopen a new one to keep it more focus as a lot have changed.
